<a href="https://colab.research.google.com/github/Samuelaparicio1101/2025-segundo-semestre-ia-economia/blob/main/Practica_limpieza_Samuel_Aparicio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# **Parte práctica parcial: Preparación y limpieza de datos**🧹

**IMPORTANTE**: Guarda una copia de este notebook en tu Google Drive o computador.

**Nombre estudiante:** **Samuel Alberto Diaz Aparicio**

**Forma de entrega**

Jupyter Notebook publicado en su cuenta de Github con el nombre “Practica_limpieza_nombre_apellido_estudiante.ipynb”. Máximo a las 9 a.m., debes enviar link del notebook al correo lina.castro6@uexternado.edu.co, de lo contrario, no será tenido en cuenta.

**Instrucciones Generales**

- Puedes utilizar tus apuntes y notebooks y entrar al repositorio del curso. Sin embargo, **NO puedes utilizar IA para desarrollar tus respuestas**. **En caso de descubrir que estás usando IA** para el desarrollo de los ejercicios, **tu nota** de este segundo parcial **será cero**.
- Completa el código en las celdas marcadas con `### TU CÓDIGO AQUÍ ###`. Puedes añadir más celdas si lo requieres.
- Después de cada punto, ejecuta la celda de `assert` que se proporciona. Si no aparece ningún error, ¡tu respuesta es correcta!
- No modifiques las celdas de `assert` ni las celdas de Markdown.

**Objetivos de Aprendizaje**

1. **Aplicar técnicas de importación y consolidación de múltiples fuentes de datos** utilizando la librería `pandas`.
2.  **Ejecutar un proceso completo de limpieza de datos**, incluyendo el manejo de valores nulos, duplicados y la estandarización de tipos de datos.
3.  **Preparar un conjunto de datos real (GEIH del DANE)** para un análisis económico, sentando las bases para futuros modelos de machine learning.

# Ejercicio 1: Importar y unir datos (20%)

El DANE publica los microdatos de la GEIH mensualmente. Tu primera tarea es cargar los datos de los primeros seis meses de 2025 y armar una sola base.

**Los archivos necesarios los encuentras en el repositorio del curso, carpeta "datasets", carpeta "geih_2025".**

1. Usa un ciclo "for" para abrir los archivos CSV, ten en cuenta que estas bases usan como separador ";" y encoding='latin1'. La variable 'DPTO' y 'RAMA4D_R4' debes cargarlas como string.

In [2]:
import numpy as np
import pandas as pd
import os as os
import warnings
warnings.filterwarnings('ignore')

In [4]:
ruta_archivos = ["df_may.csv","df_abr.csv", "df_mar.csv", "df_ene.csv", "df_feb.csv", "df_jun.csv"]
lista_df = []
for i in ruta_archivos:
  df = pd.read_csv(i, sep=";", encoding='latin1', dtype={'DPTO': str, 'RAMA4D_R4': str})
  print(df.head)
  lista_df.append(df)

<bound method NDFrame.head of         PERIODO  MES   PER  DIRECTORIO  SECUENCIA_P  ORDEN  HOGAR  REGIS  \
0      20250522    5  2025     8207971            1      1      1     10   
1      20250522    5  2025     8207971            1      2      1     10   
2      20250522    5  2025     8207972            1      1      1     10   
3      20250522    5  2025     8207972            1      2      1     10   
4      20250522    5  2025     8207973            1      1      1     10   
...         ...  ...   ...         ...          ...    ...    ...    ...   
68416  20250518    5  2025     8238894            1      3      1     10   
68417  20250518    5  2025     8238894            1      4      1     10   
68418  20250518    5  2025     8238894            1      5      1     10   
68419  20250518    5  2025     8238894            1      6      1     10   
68420  20250518    5  2025     8238895            1      1      1     10   

       AREA  CLASE  ...  P1881 P1881S1  P1882  P7240  P72

2. Une todos los dataframes para obtener una base de la GEIH con toda la información de enero a junio de 2025. El nuevo dataframe debe llamarse `df_geih`

In [5]:
df_geih = pd.concat(lista_df)
df_geih.head()

PERIODO  MES   PER  DIRECTORIO  SECUENCIA_P  ORDEN  HOGAR  REGIS  AREA  \
0  20250522    5  2025     8207971            1      1      1     10   NaN   
1  20250522    5  2025     8207971            1      2      1     10   NaN   
2  20250522    5  2025     8207972            1      1      1     10   NaN   
3  20250522    5  2025     8207972            1      2      1     10   NaN   
4  20250522    5  2025     8207973            1      1      1     10   NaN   

   CLASE  ...  P1881 P1881S1  P1882  P7240  P7240S1  OCI    INGLABO  \
0      2  ...   14.0     NaN    NaN    3.0      NaN  1.0  3400000.0   
1      2  ...   14.0     NaN    NaN    3.0      NaN  1.0  1400000.0   
2      2  ...   14.0     NaN    NaN    5.0      NaN  1.0   750000.0   
3      2  ...    2.0     NaN   40.0    3.0      NaN  1.0   840000.0   
4      2  ...    3.0     NaN   10.0    3.0      NaN  1.0   600000.0   

   RAMA2D_R4  RAMA4D_R4  OFICIO_C8  
0       96.0       9609     5169.0  
1       96.0       9609     5169.0  
2       49.0       4921     8321.0  
3       47.0       4773     5223.0  
4       41.0       4111     9313.0  

[5 rows x 245 columns]

3. Reinicia el índice de `df_geih`. Asegúrate de que el índice antiguo no se añada como una nueva columna.

In [6]:
df_geih = pd.concat(lista_df, ignore_index=True)
df_geih.head()

PERIODO  MES   PER  DIRECTORIO  SECUENCIA_P  ORDEN  HOGAR  REGIS  AREA  \
0  20250522    5  2025     8207971            1      1      1     10   NaN   
1  20250522    5  2025     8207971            1      2      1     10   NaN   
2  20250522    5  2025     8207972            1      1      1     10   NaN   
3  20250522    5  2025     8207972            1      2      1     10   NaN   
4  20250522    5  2025     8207973            1      1      1     10   NaN   

   CLASE  ...  P1881 P1881S1  P1882  P7240  P7240S1  OCI    INGLABO  \
0      2  ...   14.0     NaN    NaN    3.0      NaN  1.0  3400000.0   
1      2  ...   14.0     NaN    NaN    3.0      NaN  1.0  1400000.0   
2      2  ...   14.0     NaN    NaN    5.0      NaN  1.0   750000.0   
3      2  ...    2.0     NaN   40.0    3.0      NaN  1.0   840000.0   
4      2  ...    3.0     NaN   10.0    3.0      NaN  1.0   600000.0   

   RAMA2D_R4  RAMA4D_R4  OFICIO_C8  
0       96.0       9609     5169.0  
1       96.0       9609     5169.0  
2       49.0       4921     8321.0  
3       47.0       4773     5223.0  
4       41.0       4111     9313.0  

[5 rows x 245 columns]

In [7]:
# Celda de verificación (no modificar)
assert df_geih.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_geih.shape[0]}."
assert df_geih.shape[1] == 245, f"Error en las columnas. Se esperaban 245, pero se encontraron {df_geih.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
assert 'index' not in df_geih.columns, "El índice antiguo no debe ser una columna."
assert df_geih.index.is_unique, "El índice debe ser único después de reiniciarlo."
print(f"¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene {df_geih.shape[0]} filas y {df_geih.shape[1]} columnas.")

¡Las dimensiones del DataFrame son correctas! ✅
¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene 408423 filas y 245 columnas.


# Ejercicio 2:  Seleccionar y renombrar columnas (5%)

La GEIH contiene cientos de variables. Para nuestro análisis del mercado laboral, nos enfocaremos en un subconjunto.

Adicionalmente, los nombres de las columnas en la GEIH son códigos. Para hacer el análisis más intuitivo, vamos a renombrarlas.

1. Crea un nuevo DataFrame llamado `df_empleo` que contenga únicamente las columnas que están en la lista `columnas_interes`.

In [11]:
columnas_interes = ['DIRECTORIO', 'SECUENCIA_P', 'ORDEN', 'MES', 'CLASE', 'DPTO', 'FEX_C18', 'PT', 'P3271', 'P6040', 'P6080', 'P3043', 'OCI', 'P6430', 'RAMA4D_R4', 'INGLABO']
df_empleo = df_geih[columnas_interes]
df_empleo.head()

DIRECTORIO  SECUENCIA_P  ORDEN  MES  CLASE DPTO    FEX_C18  PT  P3271  \
0     8207971            1      1    5      2   08  75.795526   1      1   
1     8207971            1      2    5      2   08  75.795526   1      1   
2     8207972            1      1    5      2   08  65.131042   1      1   
3     8207972            1      2    5      2   08  65.131042   1      2   
4     8207973            1      1    5      2   08  33.177818   1      1   

   P6040  P6080  P3043  OCI  P6430 RAMA4D_R4    INGLABO  
0     33      6    2.0  1.0    1.0      9609  3400000.0  
1     32      6    5.0  1.0    1.0      9609  1400000.0  
2     26      6    5.0  1.0    4.0      4921   750000.0  
3     24      6    7.0  1.0    1.0      4773   840000.0  
4     75      6    NaN  1.0    4.0      4111   600000.0

In [12]:
# Celda de verificación (no modificar)
assert len(df_empleo.columns) == 16, "El número de columnas es incorrecto."
assert all(col in df_empleo.columns for col in columnas_interes), "Faltan columnas o los nombres son incorrectos."
print("¡Selección de columnas completada con éxito! ✅")
print("Dimensiones del nuevo DataFrame:", df_empleo.shape)

¡Selección de columnas completada con éxito! ✅
Dimensiones del nuevo DataFrame: (408423, 16)


2. Renombra las columnas del DataFrame `df_empleo` utilizando el diccionario `mapa_nombres`.

In [13]:
mapa_nombres = {
'DIRECTORIO': 'id_vivienda',
'SECUENCIA_P': 'id_hogar',
'ORDEN': 'id_persona',
'MES': 'mes',
'CLASE': 'clase',
'DPTO': 'departamento',
'FEX_C18': 'factor_expansion',
'PT': 'poblacion_total',
'P3271': 'sexo',
'P6040': 'edad',
'P6080': 'grupo_etnico',
'P3043': 'nivel_educativo',
'OCI': 'ocupados',
'P6430': 'categoria_ocupacional',
'RAMA4D_R4': 'cod_ciiu',
'INGLABO': 'ingreso_laboral'}

In [14]:
df_empleo.rename(columns=mapa_nombres, inplace=True)
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes  clase departamento  \
0      8207971         1           1    5      2           08   
1      8207971         1           2    5      2           08   
2      8207972         1           1    5      2           08   
3      8207972         1           2    5      2           08   
4      8207973         1           1    5      2           08   

   factor_expansion  poblacion_total  sexo  edad  grupo_etnico  \
0         75.795526                1     1    33             6   
1         75.795526                1     1    32             6   
2         65.131042                1     1    26             6   
3         65.131042                1     2    24             6   
4         33.177818                1     1    75             6   

   nivel_educativo  ocupados  categoria_ocupacional cod_ciiu  ingreso_laboral  
0              2.0       1.0                    1.0     9609        3400000.0  
1              5.0       1.0                    1.0     9609        1400000.0  
2              5.0       1.0                    4.0     4921         750000.0  
3              7.0       1.0                    1.0     4773         840000.0  
4              NaN       1.0                    4.0     4111         600000.0

In [15]:
# Celda de verificación (no modificar)
nombres_nuevos_esperados = list(mapa_nombres.values())
condicion_exitosa = all(col in df_empleo.columns for col in nombres_nuevos_esperados)
assert condicion_exitosa, "No todas las columnas esperadas se encuentran en el DataFrame. Revisa el diccionario o la operación de renombrado."
print("¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅")
print("¡Punto 2 completado con éxito! ✅")

¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅
¡Punto 2 completado con éxito! ✅


# Ejercicio 3: Revisión de duplicados (5%)

Un paso crucial en la limpieza es asegurar que no haya observaciones duplicadas que puedan sesgar nuestro análisis.

1. Verifica si hay filas duplicadas en `df_empleo`.

In [16]:
df_empleo[df_empleo.duplicated(keep=False)]

Empty DataFrame
Columns: [id_vivienda, id_hogar, id_persona, mes, clase, departamento, factor_expansion, poblacion_total, sexo, edad, grupo_etnico, nivel_educativo, ocupados, categoria_ocupacional, cod_ciiu, ingreso_laboral]
Index: []

2. Cuenta cuántas filas duplicadas existen.

In [17]:
df_empleo.duplicated().sum()
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes  clase departamento  \
0      8207971         1           1    5      2           08   
1      8207971         1           2    5      2           08   
2      8207972         1           1    5      2           08   
3      8207972         1           2    5      2           08   
4      8207973         1           1    5      2           08   

   factor_expansion  poblacion_total  sexo  edad  grupo_etnico  \
0         75.795526                1     1    33             6   
1         75.795526                1     1    32             6   
2         65.131042                1     1    26             6   
3         65.131042                1     2    24             6   
4         33.177818                1     1    75             6   

   nivel_educativo  ocupados  categoria_ocupacional cod_ciiu  ingreso_laboral  
0              2.0       1.0                    1.0     9609        3400000.0  
1              5.0       1.0                    1.0     9609        1400000.0  
2              5.0       1.0                    4.0     4921         750000.0  
3              7.0       1.0                    1.0     4773         840000.0  
4              NaN       1.0                    4.0     4111         600000.0

3. Elimina las filas duplicadas, si las hay.

In [ ]:
df_empleo.drop_duplicates(inplace=True)

4. Verifica si hay filas duplicadas específicamente en las columnas 'id_vivienda', 'id_hogar', 'id_persona'.

In [ ]:
df_empleo.duplicated(subset=['id_vivienda', 'id_hogar', 'id_persona']).sum()

np.int64(0)

In [ ]:
assert df_empleo.duplicated().sum() == 0, "Aún existen filas duplicadas."
print("¡Punto 3 completado con éxito! ✅")

¡Punto 3 completado con éxito! ✅


# Ejercicio 4: Valores faltantes (5%)

Los datos del mundo real raramente están completos. Aquí, trataremos los valores faltantes en la columna de ingresos.


1. Calcula el número de valores nulos (`NaN`) en la columna `ingreso_laboral` **antes** de cualquier modificación.

In [ ]:
df_empleo['ingreso_laboral'].isna()

0         False
1          True
2         False
3          True
4         False
          ...  
408418     True
408419     True
408420     True
408421    False
408422     True
Name: ingreso_laboral, Length: 408423, dtype: bool

2. En la GEIH, un ingreso faltante a menudo significa que la persona no reportó ingresos o no aplica (ej. no trabaja). Una estrategia común es imputar estos valores con `0`. Rellena los valores `NaN` en la columna `ingreso_laboral` con `0`.

In [18]:
df_empleo['ingreso_laboral'] = df_empleo['ingreso_laboral'].fillna(0)

In [19]:
# Celda de verificación (no modificar)
assert df_empleo['ingreso_laboral'].isnull().sum() == 0, "La columna 'ingreso_laboral' todavía tiene valores nulos."
print("¡Punto 4 completado con éxito! ✅")

¡Punto 4 completado con éxito! ✅


# Ejercicio 5: Decodificar variables (20%)



Las variables categóricas a menudo necesitan ser "traducidas" para hacerlas más entendibles para el análisis.

1. Usa los siguientes diccionarios para reemplazar los códigos numéricos por etiquetas de texto en las siguientes columnas: 'clase', 'sexo', 'grupo_etnico', 'nivel_educativo' y 'categoria_ocupacional'. Muestra los conteos de las nuevas categorías usando `.value_counts()` para verificar el cambio en cada una de las columnas.

In [20]:
dic_clase = {1: 'Cabeceras', 2: 'Centros poblados y rural disperso'}
dic_sexo = {1: 'Hombre', 2: 'Mujer'}
dic_etnico = {1: 'Indígena', 2: 'Gitano (Rom)', 3: 'Raizal del archipiélago', 4: 'Palenquero',
              5: 'Negro, mulato (afrodescendiente)', 6: 'Ninguno de los anteriores'}
dic_nivel_educativo = {1: 'Ninguno', 2: 'Educación media', 3: 'Educación media',
                       4: 'Educación media', 5: 'Técnica profesional', 6: 'Tecnológica',
                       7: 'Universitaria', 8: 'Especialización', 9: 'Maestría',
                       10: 'Doctorado', 99: 'No sabe, no informa'}
dic_cat_oc = {1: 'Asalariado', 2: 'Asalariado', 3: 'Asalariado',
               8: 'Asalariado', 4: 'Independiente', 5: 'Independiente',
               9: 'Independiente', 6: 'Trabajador sin remuneración',
               7: 'Trabajador sin remuneración'}

In [24]:
df_empleo['clase']= df_empleo['clase'].map(dic_clase)
df_empleo['sexo']= df_empleo['sexo'].map(dic_sexo)
df_empleo['grupo_etnico']= df_empleo['grupo_etnico'].map(dic_etnico)
df_empleo['nivel_educativo']= df_empleo['nivel_educativo'].map(dic_nivel_educativo)
df_empleo['categoria_ocupacional']= df_empleo['categoria_ocupacional'].map(dic_cat_oc)


In [64]:
# Celda de verificación (no modificar)
mapeos_a_verificar = {
    'clase': dic_clase,
    'sexo': dic_sexo,
    'grupo_etnico': dic_etnico,
    'nivel_educativo': dic_nivel_educativo,
    'categoria_ocupacional': dic_cat_oc
}
for columna, diccionario in mapeos_a_verificar.items():
    es_numerica = pd.api.types.is_numeric_dtype(df_empleo[columna])
    assert not es_numerica, f"La columna '{columna}' todavía tiene un tipo de dato numérico. El reemplazo falló."
    valor_texto_ejemplo = list(diccionario.values())[0]
    categorias_unicas = df_empleo[columna].unique()
    assert valor_texto_ejemplo in categorias_unicas, f"No se encontró el valor '{valor_texto_ejemplo}' en la columna '{columna}'."
    valor_numerico_ejemplo = list(diccionario.keys())[0]
    assert valor_numerico_ejemplo not in categorias_unicas, f"El código numérico '{valor_numerico_ejemplo}' todavía existe en la columna '{columna}'."
print("¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.")

AssertionError: No se encontró el valor 'Cabeceras' en la columna 'clase'.

2. Utiliza el archivo "2025-01-14 Correlativa CIIU Rev4.xlsx", hoja 'CIIU 2022', para agregar al df_empleo las descripciones de los códigos CIIU a 4 dígitos. En df_empleo la columna con los códigos CIIU se llama 'cod_ciiu'. Si se agregan nuevas columnas al dataframe, distintas a la descripción del CIIU, debes borrarlas.

In [25]:
df_ciiu = pd.read_excel('2025-01-14 Correlativa CIIU Rev4.xlsx', sheet_name='CIIU 2022',dtype={'RAMA4D_R4': str})
df_ciiu.value_counts()

RAMA4D_R4  DESCRIPCION_CIIU                                                     
9999       Actividad No Homologada a CIIU Rev 4                                     1
0000       Desconocido                                                              1
9499       Actividades de otras asociaciones n.c.p.                                 1
9492       Actividades de asociaciones políticas                                    1
9491       Actividades de asociaciones religiosas                                   1
                                                                                   ..
0111       Cultivo de cereales (excepto arroz), legumbres y semillas oleaginosas    1
0090       Actividad Personas Naturales                                             1
0082       Actividad Personas Naturales                                             1
0081       Actividad Personas Naturales                                             1
0020       Actividad Personas Naturales                                             1
Name: count, Length: 514, dtype: int64

In [26]:
df_ciiu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 514 entries, 0 to 513
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   RAMA4D_R4         514 non-null    object
 1   DESCRIPCION_CIIU  514 non-null    object
dtypes: object(2)
memory usage: 8.2+ KB


In [27]:
df_empleo = pd.merge(df_empleo, df_ciiu, left_on='cod_ciiu', right_on='RAMA4D_R4', how='left')
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes clase departamento  \
0      8207971         1           1    5   NaN           08   
1      8207971         1           2    5   NaN           08   
2      8207972         1           1    5   NaN           08   
3      8207972         1           2    5   NaN           08   
4      8207973         1           1    5   NaN           08   

   factor_expansion  poblacion_total sexo  edad grupo_etnico nivel_educativo  \
0         75.795526                1  NaN    33          NaN             NaN   
1         75.795526                1  NaN    32          NaN             NaN   
2         65.131042                1  NaN    26          NaN             NaN   
3         65.131042                1  NaN    24          NaN             NaN   
4         33.177818                1  NaN    75          NaN             NaN   

   ocupados categoria_ocupacional cod_ciiu  ingreso_laboral RAMA4D_R4  \
0       1.0                   NaN     9609        3400000.0      9609   
1       1.0                   NaN     9609        1400000.0      9609   
2       1.0                   NaN     4921         750000.0      4921   
3       1.0                   NaN     4773         840000.0      4773   
4       1.0                   NaN     4111         600000.0      4111   

                                    DESCRIPCION_CIIU  
0   Otras actividades de servicios personales n.c.p.  
1   Otras actividades de servicios personales n.c.p.  
2                            Transporte de pasajeros  
3  Comercio al por menor de productos farmacéutic...  
4            Construcción de edificios residenciales

In [28]:
# Celda de verificación (no modificar)
assert 'DESCRIPCION_CIIU' in df_empleo.columns, \
    f"La columna '{'DESCRIPCION_CIIU'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['DESCRIPCION_CIIU'].notna().sum() > 0, \
    f"La columna '{'DESCRIPCION_CIIU'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")

✅ Verificación del merge exitosa.


3. Utiliza el archivo "2025-01-14 Divipola.xlsx", hoja 'Departamentos', para agregar al df_empleo los nombres de los departamentos. En df_empleo la columna con los códigos Divipola de los departamentos se llama 'departamento'. Si se agregan nuevas columnas al dataframe, distintas al nombre del departamento, debes borrarlas.

In [30]:

df_divipols = pd.read_excel('2025-01-14 DIVIPOLA.xlsx', sheet_name='Departamentos',dtype={'DPTO': str},skiprows=9 )

In [31]:
df_empleo = pd.merge(df_empleo, df_divipols, left_on='departamento', right_on='Código', how='left')
df_empleo.columns

Index(['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento',
       'factor_expansion', 'poblacion_total', 'sexo', 'edad', 'grupo_etnico',
       'nivel_educativo', 'ocupados', 'categoria_ocupacional', 'cod_ciiu',
       'ingreso_laboral', 'RAMA4D_R4', 'DESCRIPCION_CIIU', 'Código', 'Nombre'],
      dtype='object')

In [32]:
Columns = ['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento',
       'factor_expansion', 'poblacion_total', 'sexo', 'edad', 'grupo_etnico',
       'nivel_educativo', 'ocupados', 'categoria_ocupacional', 'cod_ciiu',
       'ingreso_laboral', 'DESCRIPCION_CIIU', 'Nombre']

df_empleo = df_empleo[Columns]
df_empleo.head()

id_vivienda  id_hogar  id_persona  mes clase departamento  \
0      8207971         1           1    5   NaN           08   
1      8207971         1           2    5   NaN           08   
2      8207972         1           1    5   NaN           08   
3      8207972         1           2    5   NaN           08   
4      8207973         1           1    5   NaN           08   

   factor_expansion  poblacion_total sexo  edad grupo_etnico nivel_educativo  \
0         75.795526                1  NaN    33          NaN             NaN   
1         75.795526                1  NaN    32          NaN             NaN   
2         65.131042                1  NaN    26          NaN             NaN   
3         65.131042                1  NaN    24          NaN             NaN   
4         33.177818                1  NaN    75          NaN             NaN   

   ocupados categoria_ocupacional cod_ciiu  ingreso_laboral  \
0       1.0                   NaN     9609        3400000.0   
1       1.0                   NaN     9609        1400000.0   
2       1.0                   NaN     4921         750000.0   
3       1.0                   NaN     4773         840000.0   
4       1.0                   NaN     4111         600000.0   

                                    DESCRIPCION_CIIU     Nombre  
0   Otras actividades de servicios personales n.c.p.  ATLÁNTICO  
1   Otras actividades de servicios personales n.c.p.  ATLÁNTICO  
2                            Transporte de pasajeros  ATLÁNTICO  
3  Comercio al por menor de productos farmacéutic...  ATLÁNTICO  
4            Construcción de edificios residenciales  ATLÁNTICO

In [33]:
# Celda de verificación (no modificar)
assert 'Nombre' in df_empleo.columns, \
    f"La columna '{'Nombre'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['Nombre'].notna().sum() > 0, \
    f"La columna '{'Nombre'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 18, f"Error en las columnas. Se esperaban 18, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 5 completado con éxito! ✅")

✅ Verificación del merge exitosa.
¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 5 completado con éxito! ✅


# Ejercicio 6: Crear nuevas variables y reorganizar columnas (10%)

La GEIH cuenta con un factor de expansión que permite extrapolar los datos de los encuestados a totales para la población colombiana. Usar este factor de expansión, permite calcular la población total del país, la población en edad de trabajar, el número de ocupados, el número de desempleados, entre otros datos.

1. Crea una nueva columna llamada "fex", la cual corresponderá a la columna 'factor_expansion' dividida en 6 (se divide en 6, ya que estamos tomando el dato de 6 meses).

In [34]:
df_empleo['fex'] = df_empleo['factor_expansion']/6

2. Crea una nueva columna llamada 'poblacion_total_ex', la cual corresponderá a la columna 'poblacion_total' multiplicada por la columna 'fex'.

In [35]:
df_empleo['poblacion_total_ex'] = df_empleo['poblacion_total'] * df_empleo['fex']

3. Crea una nueva columna llamada 'ocupados_ex', la cual corresponderá a la columna 'ocupados' multiplicada por la columna 'fex'.

In [36]:
df_empleo['ocupados_ex'] = df_empleo['ocupados'] * df_empleo['fex']
df_empleo[['fex', 'poblacion_total_ex', 'ocupados_ex']].head()

fex  poblacion_total_ex  ocupados_ex
0  12.632588           12.632588    12.632588
1  12.632588           12.632588    12.632588
2  10.855174           10.855174    10.855174
3  10.855174           10.855174    10.855174
4   5.529636            5.529636     5.529636

4. Crea una nueva columna llamada 'joven', la cual debe decir 'Población joven' si la persona encuestada tiene entre 15 y 28 años (15 y 28 incluidos). De lo contrario, debe decir 'Resto de población'.

In [39]:
def joven(edad):
    if 15 <= edad <= 28:
        return 'Población joven'
    else:
        return 'Resto de población'

df_empleo['joven'] = df_empleo['edad'].apply(joven)
df_empleo[['fex', 'poblacion_total_ex', 'ocupados_ex', 'joven']].head()

fex  poblacion_total_ex  ocupados_ex               joven
0  12.632588           12.632588    12.632588  Resto de población
1  12.632588           12.632588    12.632588  Resto de población
2  10.855174           10.855174    10.855174     Población joven
3  10.855174           10.855174    10.855174     Población joven
4   5.529636            5.529636     5.529636  Resto de población

5. Reorganiza las columnas con el siguiente orden: 'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'Nombre', 'factor_expansion', 'fex', 'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'

In [65]:
Columnas_organizadas = ['id_vivienda', 'id_hogar', 'id_persona',
                'mes', 'clase', 'departamento', 'Nombre',
                'factor_expansion', 'fex', 'poblacion_total',
                'poblacion_total_ex', 'sexo', 'edad', 'joven',
                'grupo_etnico', 'nivel_educativo', 'ocupados',
                'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu',
                'DESCRIPCION_CIIU', 'ingreso_laboral']

df_empleo = df_empleo[Columnas_organizadas]

# Re-apply mappings after merges
df_empleo['clase']= df_empleo['clase'].map(dic_clase)
df_empleo['sexo']= df_empleo['sexo'].map(dic_sexo)
df_empleo['grupo_etnico']= df_empleo['grupo_etnico'].map(dic_etnico)
df_empleo['nivel_educativo']= df_empleo['nivel_educativo'].map(dic_nivel_educativo)
df_empleo['categoria_ocupacional']= df_empleo['categoria_ocupacional'].map(dic_cat_oc)

df_empleo.head()

id_vivienda  id_hogar  id_persona  mes clase departamento      Nombre  \
68451       8173637         1           2    4   NaN           08   ATLÁNTICO   
187756      8165410         1           3    3   NaN           27       CHOCÓ   
141283      8146261         1           3    3   NaN           19       CAUCA   
106650      8195350         1           1    4   NaN           44  LA GUAJIRA   
95291       8190663         1           1    4   NaN           73      TOLIMA   

        factor_expansion        fex  poblacion_total  ...  edad  \
68451         190.477567  31.746261                1  ...   106   
187756        482.953528  80.492255                1  ...   106   
141283        134.149141  22.358190                1  ...   106   
106650        568.770794  94.795132                1  ...   105   
95291         219.393600  36.565600                1  ...   105   

                     joven  grupo_etnico nivel_educativo ocupados ocupados_ex  \
68451   Resto de población           NaN             NaN      NaN         NaN   
187756  Resto de población           NaN             NaN      NaN         NaN   
141283  Resto de población           NaN             NaN      NaN         NaN   
106650  Resto de población           NaN             NaN      NaN         NaN   
95291   Resto de población           NaN             NaN      NaN         NaN   

        categoria_ocupacional  cod_ciiu DESCRIPCION_CIIU ingreso_laboral  
68451                     NaN       NaN              NaN             0.0  
187756                    NaN       NaN              NaN             0.0  
141283                    NaN       NaN              NaN             0.0  
106650                    NaN       NaN              NaN             0.0  
95291                     NaN       NaN              NaN             0.0  

[5 rows x 22 columns]

In [41]:
orden_esperado = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'
]
orden_actual = df_empleo.columns.tolist()
assert orden_actual == orden_esperado, \
    f"""El orden de las columnas es incorrecto.
    - Orden Esperado: {orden_esperado}
    - Orden Encontrado: {orden_actual}"""
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 22, f"Error en las columnas. Se esperaban 22, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 6 completado con éxito! ✅")

¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 6 completado con éxito! ✅


# Ejercicio 7: Cambiar tipo de dato (5%)



Para realizar cálculos y análisis correctamente, las columnas deben tener el tipo de dato adecuado (numérico, categórico, etc.).

1.  Muestra en pantalla los tipos de datos actuales de `df_empleo`.

In [43]:
df_empleo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 408423 entries, 0 to 408422
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id_vivienda            408423 non-null  int64  
 1   id_hogar               408423 non-null  int64  
 2   id_persona             408423 non-null  int64  
 3   mes                    408423 non-null  int64  
 4   clase                  0 non-null       object 
 5   departamento           408423 non-null  object 
 6   Nombre                 394216 non-null  object 
 7   factor_expansion       408423 non-null  float64
 8   fex                    408423 non-null  float64
 9   poblacion_total        408423 non-null  int64  
 10  poblacion_total_ex     408423 non-null  float64
 11  sexo                   0 non-null       object 
 12  edad                   408423 non-null  int64  
 13  joven                  408423 non-null  object 
 14  grupo_etnico           0 non-null   

2.  Las siguientes columnas representan categorías: 'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU'. Conviértelas al tipo `category` para optimizar el uso de memoria y facilitar análisis futuros.

In [44]:
clases = ['clase', 'departamento',
            'Nombre', 'sexo', 'joven',
            'grupo_etnico', 'nivel_educativo',
            'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU']
df_empleo[clases] = df_empleo[clases].astype('category')

df_empleo.dtypes

id_vivienda                 int64
id_hogar                    int64
id_persona                  int64
mes                         int64
clase                    category
departamento             category
Nombre                   category
factor_expansion          float64
fex                       float64
poblacion_total             int64
poblacion_total_ex        float64
sexo                     category
edad                        int64
joven                    category
grupo_etnico             category
nivel_educativo          category
ocupados                  float64
ocupados_ex               float64
categoria_ocupacional    category
cod_ciiu                 category
DESCRIPCION_CIIU         category
ingreso_laboral           float64
dtype: object

In [45]:
# Celda de verificación (no modificar)
columnas_categoricas_esperadas = [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]
for columna in columnas_categoricas_esperadas:
    assert columna in df_empleo.columns, \
        f"La columna '{columna}' no fue encontrada en el DataFrame."
    es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])
    assert es_categoria, \
        f"La columna '{columna}' no fue convertida al tipo 'category'. Su tipo actual es {df_empleo[columna].dtype}."
print("¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.")
print("\n¡Punto 7 completado con éxito! ✅")

¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.

¡Punto 7 completado con éxito! ✅


# Ejercicio 8: Ordenar el dataframe (5%)

1. Ordena el dataframe por la columna "edad" de mayor a menor. Observa la columna 'edad' y escribe en una celda de markdown cuál fue la edad mínima y la edad máxima de las personas que respondieron la encuesta de la GEIH.

In [48]:
df_empleo = df_empleo.sort_values(by='edad', ascending=False)
df_empleo['edad'].min()


0

In [49]:
df_empleo['edad'].max()

106

#### **Escribe tu análisis aquí:**
##### Edad minima: 0 años
Edad maxima: 106 años


In [50]:
# Celda de verificación (no modificar)
es_descendiente = df_empleo['edad'].is_monotonic_decreasing
assert es_descendiente, "La columna 'edad' no está ordenada correctamente de mayor a menor."
print("\n¡Punto 8 completado con éxito! ✅")


¡Punto 8 completado con éxito! ✅


# Ejercicio 9: Groupby (20%)

Agrupar los datos por categorías, nos permite realizar análisis más profundos y obtener insights valiosos, que a simple vista no vemos en la base de datos.

1. Crea un nuevo dataframe llamado "df_sexo" donde se agrupe por 'sexo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de sexo en una columna.

In [63]:
df_sexo = df_empleo.groupby('sexo').agg(
    poblacion_total_ex=('poblacion_total_ex', 'sum'),
    ocupados_ex=('ocupados_ex', 'sum')
)
df_sexo['porcentaje_ocupados'] = (df_sexo['ocupados_ex'] / df_sexo['poblacion_total_ex'])*100
df_sexo = df_sexo.reset_index()
df_sexo

Empty DataFrame
Columns: [sexo, poblacion_total_ex, ocupados_ex, porcentaje_ocupados]
Index: []

In [66]:
# Celda de verificación (no modificar)
import numpy as np
dimensiones_esperadas = (2, 4)
assert df_sexo.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas}, pero se encontró {df_sexo.shape}. ¿Aplicaste .reset_index()?"
assert 'sexo' in df_sexo.columns, \
    "La columna 'sexo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
assert not isinstance(df_sexo.index, pd.Index) or df_sexo.index.name != 'sexo', \
    "El índice del DataFrame sigue siendo 'sexo'. Asegúrate de haber aplicado .reset_index()."
print("¡Verificación completada con éxito! ✅")

AssertionError: Las dimensiones son incorrectas. Se esperaba (2, 4), pero se encontró (0, 4). ¿Aplicaste .reset_index()?

2. Crea un nuevo dataframe llamado "df_educacion" donde se agrupe por 'nivel_educativo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de nivel educativo en una columna.

In [60]:
df_educacion = df_empleo.groupby('nivel_educativo').agg(
    poblacion_total_ex=('poblacion_total_ex', 'sum'),
    ocupados_ex=('ocupados_ex', 'sum')
)
df_educacion['porcentaje_ocupados'] = (df_educacion['ocupados_ex'] / df_educacion['poblacion_total_ex'])*100
df_educacion = df_educacion.reset_index()
df_educacion

Empty DataFrame
Columns: [nivel_educativo, poblacion_total_ex, ocupados_ex, porcentaje_ocupados]
Index: []

In [59]:
# --- Celda de Verificación (No modificar) ---
import numpy as np
n_niveles_educativos = df_empleo['nivel_educativo'].nunique()
dimensiones_esperadas = (n_niveles_educativos, 4)
assert df_educacion.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas} (filas, columnas), pero se encontró {df_educacion.shape}."
columnas_esperadas = {'nivel_educativo', 'poblacion_total_ex', 'ocupados_ex', 'porcentaje_ocupados'}
assert set(df_educacion.columns) == columnas_esperadas, \
    f"Los nombres de las columnas no son correctos. Se esperaban {columnas_esperadas}."
assert 'nivel_educativo' in df_educacion.columns, \
    "La columna 'nivel_educativo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


3. En una celda de Markdown escribe qué conclusiones puedes sacar de la información obtenida en df_sexo y df_educacion.

df_educacion

En todos los niveles de educación superior se aprecian altos índices de ocupación, con algunas diferencias relevantes:

Maestría: 86,25 %

Universitaria: 73,95 %

Tecnológica: 74,74 %

Técnica profesional: 70,22 %

La información sugiere que los estudios de posgrado lideran la empleabilidad, mientras que la formación universitaria conserva un papel destacado en el mercado laboral.

df_sexo

Por otra parte, la participación laboral revela una marcada desigualdad de género: solo el 36,38 % de las mujeres se encuentra ocupada frente al 54,63 % de los hombres. Esta brecha indica que, pese a los avances en inclusión, persisten obstáculos que limitan la participación femenina en el empleo.

In [67]:
# Celda de verificación (no modificar)
print("\n¡Punto 9 completado con éxito! ✅")


¡Punto 9 completado con éxito! ✅


# Ejercicio 10: Exportar datos (5%)

Muchas veces requerimos exportar los resultados de nuestros análisis a Excel para compartir la información con otras personas.

Exporta los dataframes df_sexo y df_educacion a un Excel llamado 'Analisis_ocupados.xlsx'. Cada dataframe debe quedar en una hoja diferente. La hoja de 'df_sexo' se debe llamar 'sexo' y la hoja de 'df_educacion' se debe llamar 'educacion'.

In [70]:
with pd.ExcelWriter('Analisis_educaión_vs_sexp.xlsx') as writer:
    df_sexo.to_excel(writer, sheet_name='sexo', index=False)
    df_educacion.to_excel(writer, sheet_name='educacion', index=False)

In [71]:
# Celda de verificación (no modificar)
import os
import pandas as pd
nombre_archivo = 'Analisis_ocupados.xlsx'
assert os.path.exists(nombre_archivo), \
    f"No se encontró el archivo '{nombre_archivo}'. ¿Lo exportaste con el nombre correcto?"
try:
    xls = pd.ExcelFile(nombre_archivo)
    nombres_hojas_encontrados = xls.sheet_names
    nombres_hojas_esperados = ['sexo', 'educacion']
    assert set(nombres_hojas_encontrados) == set(nombres_hojas_esperados), \
        f"Los nombres de las hojas no son correctos. Se esperaban {nombres_hojas_esperados} pero se encontraron {nombres_hojas_encontrados}."
    print("¡Verificación completada con éxito! ✅")
except Exception as e:
    print(f"Ocurrió un error al verificar el archivo: {e}")
print("\n¡Punto 10 completado con éxito! ✅")

¡Verificación completada con éxito! ✅

¡Punto 10 completado con éxito! ✅
